File: pachi_api.py
import subprocess
import psutil
import os
from flask_cors import CORS
from flask import Flask, request, jsonify

app = Flask(__name__)
CORS(app)

# Use a single global Pachi process
pachi_process = None

def get_pachi_config(rank):
    """Get Pachi configuration based on rank"""
    configs = {
        'beginner': {
            'max_games': 10,
            'max_tree_size': 100,
            'max_playouts': 100,
            # 'time': 20,        # Thời gian suy nghĩ tối đa (giây)
            # 'quick': True,    # Chế độ nhanh
            # 'nopass': True,   # Không cho phép pass
            # 'ladder': 1,      # Mức độ xử lý ladder
            # 'komi': 6.5       # Điểm komi cho quân trắng
        },
        'intermediate': {
            'max_games': 1000,
            'max_tree_size': 10000,
            'max_playouts': 10000,
            # 'time': 10,       # Thời gian suy nghĩ tối đa (giây)
            # 'quick': False,   # Chế độ bình thường
            # 'nopass': True,   # Không cho phép pass
            # 'ladder': 2,      # Mức độ xử lý ladder
            # 'komi': 6.5       # Điểm komi cho quân trắng
        },
        'pro': {
            'max_games': 10000,
            'max_tree_size': 100000,
            'max_playouts': 100000,
            # 'time': 30,       # Thời gian suy nghĩ tối đa (giây)
            # 'quick': False,   # Chế độ bình thường
            # 'nopass': True,   # Không cho phép pass
            # 'ladder': 3,      # Mức độ xử lý ladder
            # 'komi': 6.5       # Điểm komi cho quân trắng
        }
    }
    return configs.get(rank, configs['intermediate'])

def start_pachi(rank='intermediate'):
    # Start Pachi with default configuration (no extra flags)
    cmd = ['./pachi', '--gtp']
    return subprocess.Popen(
        cmd,
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        universal_newlines=True,
        bufsize=1
    )

def get_or_create_pachi(rank='intermediate'):
    global pachi_process
    if pachi_process is None or pachi_process.poll() is not None:
        print(f"(Re)starting global Pachi instance with rank {rank}")
        pachi_process = start_pachi(rank)
    return pachi_process

def send_gtp_command(cmd, rank='intermediate'):
    """Send GTP command to specific Pachi instance"""
    pachi = get_or_create_pachi(rank)
    try:
        pachi.stdin.write(cmd + '\n')
        pachi.stdin.flush()
        response_lines = []
        while True:
            line = pachi.stdout.readline()
            if not line:
                break  # EOF or process died
            line = line.strip()
            if line.startswith('=') or line.startswith('?'):
                # GTP response found
                # Remove the '=' or '?' and any leading space
                response = line[1:].strip()
                response_lines.append(response)
                break
            elif line != '':
                response_lines.append(line)
        return '\n'.join(response_lines)
    except (BrokenPipeError, IOError):
        print(f"Restarting global Pachi instance due to error")
        global pachi_process
        pachi_process = start_pachi(rank)
        return send_gtp_command(cmd, rank)

def send_multiple_commands(commands, rank='intermediate'):
    """Send multiple GTP commands to specific Pachi instance"""
    responses = {}
    for cmd in commands:
        response = send_gtp_command(cmd, rank)
        responses[cmd] = response
    return responses

@app.route('/batch', methods=['POST'])
def batch_commands():
    commands = request.json.get('commands', [])
    rank = request.json.get('rank', 'intermediate')
    if not commands:
        return jsonify({'error': 'No commands provided'}), 400
    print(f"Processing batch commands with rank {rank}")
    responses = send_multiple_commands(commands, rank)
    return jsonify({'responses': responses})

@app.route('/genmove', methods=['POST'])
def genmove():
    color = request.json.get('color', 'b')
    rank = request.json.get('rank', 'intermediate')
    response = send_gtp_command(f'genmove {color}', rank)
    return jsonify({'response': response})

@app.route('/play', methods=['POST'])
def play():
    color = request.json.get('color')
    vertex = request.json.get('vertex')
    rank = request.json.get('rank', 'intermediate')
    response = send_gtp_command(f'play {color} {vertex}', rank)
    return jsonify({'response': response})

@app.route('/boardsize', methods=['POST'])
def boardsize():
    size = request.json.get('size', 19)
    rank = request.json.get('rank', 'intermediate')
    response = send_gtp_command(f'boardsize {size}', rank)
    return jsonify({'response': response})

@app.route('/clear', methods=['POST'])
def clear_board():
    rank = request.json.get('rank', 'intermediate')
    response = send_gtp_command('clear_board', rank)
    return jsonify({'response': response})

@app.route('/quit', methods=['POST'])
def quit_pachi():
    global pachi_process
    if pachi_process is not None:
        send_gtp_command('quit')
        pachi_process = None
    return jsonify({'status': 'Pachi quit'})

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=3000, ssl_context=('cert.pem', 'key.pem'))

File: pachi_api.js
const express = require('express');
const cors = require('cors');
const { spawn } = require('child_process');
const app = express();

app.use(cors());
app.use(express.json());

// Use a single global Pachi process
let pachiProcess = null;

function getPachiConfig(rank) {
    const configs = {
        'beginner': {
            max_games: 10,
            max_tree_size: 100,
            max_playouts: 100,
            time: 20,        // Maximum thinking time (seconds)
            quick: true,     // Quick mode
            // nopass: true,  // Disable pass
            // ladder: 1,     // Ladder handling level
            // komi: 6.5      // Komi for white
        },
        'intermediate': {
            max_games: 1000,
            max_tree_size: 10000,
            max_playouts: 10000,
            time: 10,        // Maximum thinking time (seconds)
            quick: false,    // Normal mode
            // nopass: true,  // Disable pass
            // ladder: 2,     // Ladder handling level
            // komi: 6.5      // Komi for white
        },
        'pro': {
            max_games: 10000,
            max_tree_size: 100000,
            max_playouts: 100000,
            time: 30,        // Maximum thinking time (seconds)
            quick: false,    // Normal mode
            // nopass: true,  // Disable pass
            // ladder: 3,     // Ladder handling level
            // komi: 6.5      // Komi for white
        }
    };
    return configs[rank] || configs['intermediate'];
}

function startPachi(rank = 'intermediate') {
    const config = getPachiConfig(rank);
    const cmd = [
        './pachi',
        '--gtp',
        `--max-games=${config.max_games}`,
        `--max-tree-size=${config.max_tree_size}`,
        `--max-playouts=${config.max_playouts}`,
        `--time=${config.time}`,
        config.quick ? '--quick' : null,
        // config.nopass ? '--nopass' : null,
        // `--ladder=${config.ladder}`,
        // `--komi=${config.komi}`,
    ].filter(Boolean);

    return spawn(cmd[0], cmd.slice(1), {
        stdio: ['pipe', 'pipe', 'pipe']
    });
}

function getOrCreatePachi(rank = 'intermediate') {
    if (!pachiProcess || pachiProcess.killed) {
        console.log(`(Re)starting global Pachi instance with rank ${rank}`);
        pachiProcess = startPachi(rank);
    }
    return pachiProcess;
}

function sendGtpCommand(cmd, rank = 'intermediate') {
    return new Promise((resolve, reject) => {
        const pachi = getOrCreatePachi(rank);
        
        try {
            pachi.stdin.write(cmd + '\n');
            
            let responseLines = [];
            let responseHandler = (data) => {
                const line = data.toString().trim();
                if (line.startsWith('=') || line.startsWith('?')) {
                    // GTP response found
                    // Remove the '=' or '?' and any leading space
                    const response = line.slice(1).trim();
                    responseLines.push(response);
                    pachi.stdout.removeListener('data', responseHandler);
                    resolve(responseLines.join('\n'));
                } else if (line !== '') {
                    responseLines.push(line);
                }
            };

            pachi.stdout.on('data', responseHandler);

            // Handle errors
            pachi.stderr.on('data', (data) => {
                console.error(`Pachi stderr: ${data}`);
            });

            pachi.on('error', (error) => {
                console.error(`Pachi process error: ${error}`);
                pachiProcess = null;
                reject(error);
            });

        } catch (error) {
            console.error(`Error sending command to Pachi: ${error}`);
            pachiProcess = null;
            reject(error);
        }
    });
}

async function sendMultipleCommands(commands, rank = 'intermediate') {
    const responses = {};
    for (const cmd of commands) {
        try {
            const response = await sendGtpCommand(cmd, rank);
            responses[cmd] = response;
        } catch (error) {
            console.error(`Error in command ${cmd}: ${error}`);
            responses[cmd] = `Error: ${error.message}`;
        }
    }
    return responses;
}

// API Routes
app.post('/batch', async (req, res) => {
    const commands = req.body.commands || [];
    const rank = req.body.rank || 'intermediate';
    
    if (!commands.length) {
        return res.status(400).json({ error: 'No commands provided' });
    }
    
    console.log(`Processing batch commands with rank ${rank}`);
    try {
        const responses = await sendMultipleCommands(commands, rank);
        res.json({ responses });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

app.post('/genmove', async (req, res) => {
    const color = req.body.color || 'b';
    const rank = req.body.rank || 'intermediate';
    try {
        const response = await sendGtpCommand(`genmove ${color}`, rank);
        res.json({ response });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

app.post('/play', async (req, res) => {
    const { color, vertex } = req.body;
    const rank = req.body.rank || 'intermediate';
    try {
        const response = await sendGtpCommand(`play ${color} ${vertex}`, rank);
        res.json({ response });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

app.post('/boardsize', async (req, res) => {
    const size = req.body.size || 19;
    const rank = req.body.rank || 'intermediate';
    try {
        const response = await sendGtpCommand(`boardsize ${size}`, rank);
        res.json({ response });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

app.post('/clear', async (req, res) => {
    const rank = req.body.rank || 'intermediate';
    try {
        const response = await sendGtpCommand('clear_board', rank);
        res.json({ response });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

app.post('/quit', async (req, res) => {
    if (pachiProcess) {
        try {
            await sendGtpCommand('quit');
            pachiProcess = null;
        } catch (error) {
            console.error(`Error quitting Pachi: ${error}`);
        }
    }
    res.json({ status: 'Pachi quit' });
});

// Start server
const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`Pachi API server running on port ${PORT}`);
}); 